jujugui/static/gui/src/app/init.js

```diff
@@ -659,13 +693,97 @@ class GUIApp {
     @param {Function} next - Run the next route handler, if any.
   */
   _handleModelState(state, next) {
-    const modelAPI = this.modelAPI;
-    if (this.modelUUID !== state.model.uuid ||
-        (!modelAPI.get('connected') && !modelAPI.get('connecting') &&
-        !modelAPI.loading)) {
-      this._switchModelToUUID(state.model.uuid);
+    const newModelUUID = state.model.uuid;
+    if (this.activeModelUUID === newModelUUID) {
+      if (this.modelConnectionStatus === 'ready') {
+        next();
+        return;
+      } else if (this.modelConnectionStatus === 'connecting') {
+        // If we get in here it's because the application has dispatched multiple
+        // times but we do not want to connect multiple times.
+        return;
+      }
     }
-    next();
+
+    const modelURL = jujulib.generateModelURL(
+      this.applicationConfig.apiAddress, newModelUUID);
+
+    const connectionOptions = {
+      debug: true,
+      facades: [
+        // Sort facades alphabetically.
+        require('@canonical/jujulib/api/facades/application-v6.js'),
+        require('@canonical/jujulib/api/facades/all-watcher-v1.js'),
+        require('@canonical/jujulib/api/facades/charms-v2.js'),
+        require('@canonical/jujulib/api/facades/client-v1.js'),
+        require('@canonical/jujulib/api/facades/pinger-v1.js')
+      ],
+      wsclass: WebSocket,
+      bakery: this.bakery
+    };
+
+    this.activeModelUUID = newModelUUID;
+    this.modelConnectionStatus = 'connecting';
+
+    jujulib
+      .connectAndLogin(modelURL, {}, connectionOptions)
+      .then(juju => {
+        this.modelConnection = juju.conn;
+        const facades = juju.conn.facades;
+        // Setup pinger, if it's not running then the model will
+        // automatically disconnect after 1 minute.
+        facades.pinger.pingForever(30000);
+
+        this.activeModelWatcherHandle = facades.client.watch((err, result) => {
+          if (err) {
+            console.log('cannot watch model:', err);
+            process.exit(1);
+          }
+          this._processDeltas(result);
+          this.modelConnectionStatus = 'ready';
+        });
+
+      });
+  }
+
+  /**
+    Process the deltas that are returned from the megawatcher. This code
+    was moved over from the old model api handler and can be improved once we
+    remove the YUI db and replace it with maraca.
+    @param {Object} result The data result from the watch jujulib calls.
+  */
+  _processDeltas(result) {
+    const deltas = [];
+    const cmp = {
+      applicationInfo: 1,
+      relationInfo: 2,
+      unitInfo: 3,
+      machineInfo: 4,
+      annotationInfo: 5,
+      remoteapplicationInfo: 100
+    };
+    result.deltas.forEach(delta => {
+      const [kind, operation, entityInfo] = delta;
+      deltas.push([kind + 'Info', operation, entityInfo]);
+    });
+    deltas.sort((a, b) => {
+      // Sort items not in our hierarchy last.
```

Comment:
"last" or "list "?

Reply:
This was copied over and updated....I'm guessing it should be list :)

Reply:
It actually is "last" as in "to put the items, not in the list, in the last position" I'll update the comment to be clearer.

Reply:
Oh, hah, thanks.

Reply:
done